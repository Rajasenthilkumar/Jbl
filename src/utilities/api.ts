import envConfig from 'config/config';
import { getLoginData } from './auth.storage';

export interface ClientResponse<T> {
  status: number;
  data: T;
}

export interface ApiError {
  errorMessage: string;
  // biome-ignore lint/suspicious/noExplicitAny: data can be anything is not predictable
  data: any;
  status: number | null;
}

export async function api<T>(
  endpoint: string,
  { body, method, ...customConfig }: Partial<RequestInit> = {},
): Promise<ClientResponse<T>> {
  // API headers
  const headers: HeadersInit = { 'Content-Type': 'application/json' };

  // get user token from the store;
  const userData = getLoginData();
  // append the user token to the headers if it exists
  if (userData) {
    // biome-ignore lint/complexity/useLiteralKeys: <explanation>
    headers['Authorization'] = userData.token;
  }

  const config: RequestInit = {
    method: method,
    ...customConfig,
    headers: {
      ...headers,
      ...customConfig.headers,
    },
  };

  if (body) {
    config.body = JSON.stringify(body);
  }

  // biome-ignore lint/suspicious/noImplicitAnyLet: its a type assertion
  let data;
  let status: number | null = null;
  const url = endpoint.startsWith('http')
    ? endpoint
    : `${envConfig.apiUrl}${endpoint}`;
  try {
    const response = await window.fetch(url, config);
    try {
      data = await response.json();
    } catch {
      data = null;
    }
    status = response.status;

    if (response.ok) {
      return {
        status,
        data,
      };
    }
    throw new Error(response.statusText);
    // biome-ignore lint/suspicious/noExplicitAny: its a type assertion
  } catch (err: any) {
    const error: ApiError = {
      errorMessage: err.message ?? '',
      data: data ?? null,
      status: status ?? null,
    };
    return Promise.reject(error);
  }
}

// Adding the GET method
api.get = <T>(endpoint: string, customConfig: Partial<RequestInit> = {}) =>
  api<T>(endpoint, { ...customConfig, method: 'GET' });

// Adding the POST method
api.post = <T>(
  endpoint: string,
  // biome-ignore lint/suspicious/noExplicitAny: <explanation>
  body: any,
  customConfig: Partial<RequestInit> = {},
) => api<T>(endpoint, { ...customConfig, body, method: 'POST' });

// Adding the PUT method
api.put = <T>(
  endpoint: string,
  // biome-ignore lint/suspicious/noExplicitAny: <explanation>
  body: any,
  customConfig: Partial<RequestInit> = {},
) => api<T>(endpoint, { ...customConfig, body, method: 'PUT' });

// Adding the DELETE method
api.delete = <T>(
  endpoint: string,
  // biome-ignore lint/suspicious/noExplicitAny: <explanation>
  body?: any,
  customConfig: Partial<RequestInit> = {},
) => api<T>(endpoint, { ...customConfig, body, method: 'DELETE' });
