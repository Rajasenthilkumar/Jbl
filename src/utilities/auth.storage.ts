import {
  LOGIN_STORAGE_KEY,
  REMEMBER_ME_GUEST_STORAGE_KEY,
  REMEMBER_ME_HOST_STORAGE_KEY,
} from 'constants/localStorageKeys';
import { z } from 'zod';

const LoginSchema = z.object({
  token: z.string(),
  id: z.number(),
  mode: z.string(),
});

const RememberMeSchema = z.object({
  email: z.string().optional(),
  remember_me: z.boolean(),
});
const RememberGuestSchema = z.object({
  id_number: z.string().optional(),
  remember_me: z.boolean(),
});

type Login = z.infer<typeof LoginSchema>;
type RememberMe = z.infer<typeof RememberMeSchema>;
type RememberGuest = z.infer<typeof RememberGuestSchema>;

export const saveLoginDataToLocal = (value: Login) => {
  try {
    LoginSchema.parse(value);
    localStorage.setItem(LOGIN_STORAGE_KEY, JSON.stringify(value));
  } catch (error) {
    console.error('Failed to save login data:', error);
  }
};

export const saveRememberMeDataToLocal = (value: RememberMe) => {
  try {
    RememberMeSchema.parse(value);
    localStorage.setItem(REMEMBER_ME_HOST_STORAGE_KEY, JSON.stringify(value));
  } catch (error) {
    console.error('Failed to save Remember-Me data', error);
  }
};
export const getRememberMeData = (): RememberMe | null => {
  let storedValue = sessionStorage.getItem(LOGIN_STORAGE_KEY);

  if (!storedValue) {
    storedValue = localStorage.getItem(REMEMBER_ME_HOST_STORAGE_KEY);
  }

  if (storedValue) {
    try {
      const parsedValue = JSON.parse(storedValue);
      return RememberMeSchema.parse(parsedValue);
    } catch (error) {
      console.error('Failed to retrieve or validate remember-me data:', error);
      return null;
    }
  }

  return null;
};

export const saveRememberGuestDataToLocal = (value: RememberGuest) => {
  try {
    RememberGuestSchema.parse(value);
    localStorage.setItem(REMEMBER_ME_GUEST_STORAGE_KEY, JSON.stringify(value));
  } catch (error) {
    console.error('Failed to save Remember-Me data', error);
  }
};
export const getRememberGuestData = (): RememberGuest | null => {
  let storedValue = sessionStorage.getItem(LOGIN_STORAGE_KEY);

  if (!storedValue) {
    storedValue = localStorage.getItem(REMEMBER_ME_GUEST_STORAGE_KEY);
  }

  if (storedValue) {
    try {
      const parsedValue = JSON.parse(storedValue);
      return RememberGuestSchema.parse(parsedValue);
    } catch (error) {
      console.error('Failed to retrieve or validate remember-me data:', error);
      return null;
    }
  }

  return null;
};

export const getLoginData = (): Login | null => {
  let storedValue = sessionStorage.getItem(LOGIN_STORAGE_KEY);

  if (!storedValue) {
    storedValue = localStorage.getItem(LOGIN_STORAGE_KEY);
  }

  if (storedValue) {
    try {
      const parsedValue = JSON.parse(storedValue);
      return LoginSchema.parse(parsedValue);
    } catch (error) {
      console.error('Failed to retrieve or validate login data:', error);
      return null;
    }
  }

  return null;
};

export const removeLoginData = () => {
  localStorage.removeItem(LOGIN_STORAGE_KEY);
};
