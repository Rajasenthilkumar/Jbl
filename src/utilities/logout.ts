import toast from 'react-hot-toast';
import { cleanAuthInfo } from 'stores/slice/authSlice';
import { resetPropertySlice } from 'stores/slice/propertiesSlice';
import store from 'stores/store';
import { removeLoginData } from './auth.storage';

export const logout = () => {
  // clean the auth redux
  store.dispatch(cleanAuthInfo());
  // clean the property redux
  store.dispatch(resetPropertySlice());
  // clean the local storage
  removeLoginData();
  toast.success('Logout successfully', {
    duration: 1000,
  });
};
