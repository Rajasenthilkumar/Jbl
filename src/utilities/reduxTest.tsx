import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import type React from 'react';
import type { PropsWithChildren } from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { setupStore } from 'stores/setUpStore';
import type { AppStore, RootState } from 'stores/types';

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: Partial<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {},
) {
  // biome-ignore lint/complexity/noBannedTypes: its required for testing
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return (
      <MemoryRouter>
        <Provider store={store}>{children}</Provider>
      </MemoryRouter>
    );
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
