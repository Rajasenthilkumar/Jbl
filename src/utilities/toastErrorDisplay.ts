import toast from 'react-hot-toast';
import type { ApiError } from './api';

export const toastErrorDisplay = (error: unknown) => {
  const apiError: ApiError = error as ApiError;

  if (apiError.data?.error?.msg) {
    const msg = apiError.data.error.msg;

    // Check if msg is an object
    if (typeof msg === 'object' && msg !== null) {
      // Iterate over each key in the msg object and show a toast for each
      Object.entries(msg).forEach(([field, message]) => {
        toast.error(`${field}: ${message}`);
      });
    } else {
      // If it's a string or any other type, display it directly
      toast.error(msg.toString());
    }
  } else {
    if (apiError.status === 404) {
      toast.error('404 Error - Method not found...');
    } else {
      toast.error(apiError.errorMessage.toString());
    }
  }
};
