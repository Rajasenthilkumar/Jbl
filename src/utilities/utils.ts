// utils for remove characters that are not digits
export const removeNonDigits = (value: string) => {
  return value.replace(/[^\d]/g, '');
};

// 5mb
export const maxSize = 1024 * 1024 * 5;

// utils for extract file name from url eg: https://jbl-dev.s3.us-east-1.amazonaws.com/1727237811842-jbl%20logo.png return logo.png
//  handle error if url is not valid
export const extractFileName = (url: string) => {
  try {
    const urlSplit = url.split('/');
    return urlSplit[urlSplit.length - 1];
  } catch {
    return 'unknown';
  }
};
