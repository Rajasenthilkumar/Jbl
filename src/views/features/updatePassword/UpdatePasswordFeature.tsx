import type { FC } from 'react';
import toast from 'react-hot-toast';
import { useNavigate } from 'react-router-dom';
import { toastErrorDisplay } from 'utilities/toastErrorDisplay';
import UpdatePasswordForm, { type ResetPasswordFormType } from './Form';
import { useUpdatePassword } from './hooks';

type Props = {
  token: string | null;
  mode: 'Host' | 'Guest';
};

const UpdatePasswordFeature: FC<Props> = ({ token, mode }) => {
  const { updatePassword, loading } = useUpdatePassword();
  const navigate = useNavigate();

  const onFinishForm = async (updatePasswordForm: ResetPasswordFormType) => {
    try {
      if (token === null) {
        toast.error('Token is null');
        return;
      }

      const payload = {
        password: updatePasswordForm.password,
        token,
        mode,
      };
      await updatePassword(payload);
      toast.success('Password reset is successful');
      navigate('/login');
    } catch (error) {
      toastErrorDisplay(error);
    }
  };

  return (
    <div>
      <UpdatePasswordForm
        onFinish={onFinishForm}
        isSubmitting={loading}
        mode={mode}
      />
    </div>
  );
};

export default UpdatePasswordFeature;
