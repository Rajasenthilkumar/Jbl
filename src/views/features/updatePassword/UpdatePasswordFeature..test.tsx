import { renderWithProviders } from 'utilities/reduxTest';
import { describe, expect, it } from 'vitest';
import UpdatePasswordFeature from './UpdatePasswordFeature';

describe('UpdatePasswordFeature page', () => {
  const setup = () => {
    renderWithProviders(
      <UpdatePasswordFeature token={'sample'} mode={'Host'} />,
    );
  };
  it('UpdatePasswordFeature snap shot', () => {
    expect(setup()).toMatchSnapshot();
  });
});
