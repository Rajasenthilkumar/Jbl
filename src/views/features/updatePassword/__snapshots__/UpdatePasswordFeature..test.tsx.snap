// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AuthLayout page > AuthLayout snap shot 1`] = `undefined`;

exports[`UpdatePasswordFeature page > UpdatePasswordFeature snap shot 1`] = `undefined`;
